### DISCUSSION TITLE
No matching bindings found for serviceIdentifier

### DISCUSSION THREAD

**Discussion by csamarchi** [eclipse-glsp/glsp#699] [General]
Hi again, when I launch my node server and GLSP VScode extension I get this error in the debug console: 

`(node:83544) UnhandledPromiseRejectionWarning: Error: No matching bindings found for serviceIdentifier: ModuleModelState
`

<img width="1442" alt="Screen Shot 2022-07-05 at 4 35 27 PM" src="https://user-images.githubusercontent.com/42501494/177420497-fe605ad9-96ae-417e-aad8-fbad9bac0e30.png">

Meanwhile, the client extension endlessly stalls as shown below: 

<img width="951" alt="Screen Shot 2022-07-05 at 4 36 22 PM" src="https://user-images.githubusercontent.com/42501494/177420670-762f9bc1-a913-40c3-b992-6f73fed72bd0.png">

I have tried adding ModelState in my GModelDiagramModule file following this [example](https://github.com/eclipse-glsp/glsp-server-node/blob/main/examples/workflow-server/src/workflow-diagram-module.ts) like so and this does not solve it. 

`    protected bindModuleModelState(): BindingTarget<ModelState> | undefined {
        return ModuleModelState;
    }`

I was hoping someone could point me in the right direction or if you have any ideas. 

Thanks!

**Comment by tortmayr** [comment]
HI @csamarchi,

if you want  to bind a custom model state you have to override the `bindModelState()` method. I'd recommend to return a `ServiceTarget`. This ensures that both `(@inject(ModelState)` and `@inject(ModuleModleState)` resolve to the same instance:

```ts
    protected bindModelState(): BindingTarget<ModelState> {
        return { service: ModuleModelState };
    }
```

**Reply by csamarchi** [reply]
I removed the ovverride because i got the error `This member cannot have an 'override' modifier because it is not declared in the base class 'GModelDiagramModule'.ts(4113)`

**Reply by csamarchi** [reply]
So if you have a work around that error, I have to remove the override.

**Reply by csamarchi** [reply]
Is it possible to get around this error without using override?

**Comment by tortmayr** [comment]
>I removed the ovverride because i got the error This member cannot have an 'override' modifier because it is not declared in the base class 'GModelDiagramModule'.ts(4113)
>Is it possible to get around this error without using override?

Sorry, I'm afraid I can't follow. What exactly do you mean with `I removed the override`? The snippet I posted doesn't contain any override modifier and for me it compiles just fine without any errors.

>What do you mean return a ServiceTarget

Each `bind-Method` (e.g. `bindModelState()`) expects a [BindingTarget](https://github.com/eclipse-glsp/glsp-server-node/blob/8d7c5f8d3f2db01a8a973055e7f3cfe38efe01f2/packages/server-node/src/di/binding-target.ts#L74) as return value.
Typically this is just the class you want to bind eg:
```ts
    protected bindActionDispatcher(): BindingTarget<ActionDispatcher> {
        return DefaultActionDispatcher;
    }
```

However, you can also return more complex `BindingsTargets`. The following code snippet returns a `ServiceTarget`:
```ts
protected bindModelState(): BindingTarget<ModelState> {
        return { service: ModuleModelState };
    }
```
So intead of a class your return an object with  `service` property. Using a service target has the advantage that under the hood both `ModelState` and `ModuleModelState` are configured to use the same binding. So you can use either `@inject(ModelState)` or `@inject(ModuleModleState)` and will receive the same instance.

**Reply by csamarchi** [reply]
Thanks for the reply. With your solution, I am still getting errors. 

<img width="565" alt="Screen Shot 2022-07-11 at 4 19 50 PM" src="https://user-images.githubusercontent.com/42501494/178360320-44e7d600-00eb-4bc8-8570-376e42153115.png">

One that has to do with the `override`: 
`This member must have an 'override' modifier because it overrides a member in the base class 'GModelDiagramModule'`

And another:
`Property 'bindModelState' in type 'WorkflowDiagramModule' is not assignable to the same property in base type 'GModelDiagramModule'.
  Type '() => BindingTarget<ModelState> | undefined' is not assignable to type '() => BindingTarget<ModelState>'.`

**Reply by csamarchi** [reply]
Also still receiving the initial "No matching bindings" in debug console

**Reply by tortmayr** [reply]
Ah I see, looks like the correct binding is :

```ts
protected override bindModelState(): BindingTarget<ModelState> {
        return { service: ModuleModelState };
}
```
I also prepared a simple working [example](https://github.com/eclipse-glsp/glsp-server-node/commit/5cbe4636cbb582ff09602da38a7ce83840f96a47) that rebinds to a custom model state and injects the custom model state in the `CreateWorkflowNodeOperationHandler` class.

**Reply by csamarchi** [reply]
that's right, but I still am getting the binding error in my debug console for the server. Are you not getting this? It must have something to do with our model state we've created? 
<img width="1073" alt="Screen Shot 2022-07-13 at 5 13 06 PM" src="https://user-images.githubusercontent.com/42501494/178845889-db4908fb-4ffa-49a0-b2ea-85ed7d71692f.png">

**Reply by csamarchi** [reply]
are you saying I should do the same with my node-operation-handle file and inject my custom model state in order for it to work? i am still getting the above error when doing so

**Reply by tortmayr** [reply]
The error you receive is not related to your custom model state, it's about bindings for `ModuleModelIndex`. I assume this is your custom implementation of `GModelIndex`? If so how did you bind it? You have to bind the same way as you did bind the `ModelState`:

```ts
    protected override bindGModelIndex(): BindingTarget<GModelIndex> {
        return { service: ModuleModelIndex };
    }
```
I have updated the [example](https://github.com/eclipse-glsp/glsp-server-node/commit/5cbe4636cbb582ff09602da38a7ce83840f96a47) and it now also binds a custom model index.


>are you saying I should do the same with my node-operation-handle file and inject my custom model state in order for it to work? i am still getting the above error when doing so

No, not all. I simply used the `CreateWorkflowNodeOperationHandler` class as an example to test whether the binding works.

**Reply by csamarchi** [reply]
Thank you, this helped a lot
